src/ring_theory/perfection.lean

```diff
@@ -0,0 +1,422 @@
+    rw [← ring_hom.map_pow, coeff_pow_p]; assumption },
+  rw [val_aux_eq (coeff_add_ne_zero hm 1), val_aux_eq (coeff_add_ne_zero hn 1), val_aux_eq hfg],
+  rw ring_hom.map_mul at hfg ⊢, rw [mod_p.pre_val_mul hfg, mul_pow]
+end
+
+lemma val_aux_add (f g : pre_tilt K v O hv p) :
+  val_aux K v O hv p (f + g) ≤ max (val_aux K v O hv p f) (val_aux K v O hv p g) :=
+begin
+  by_cases hf : f = 0, { rw [hf, zero_add, val_aux_zero, max_eq_right], exact zero_le _ },
+  by_cases hg : g = 0, { rw [hg, add_zero, val_aux_zero, max_eq_left], exact zero_le _ },
+  by_cases hfg : f + g = 0, { rw [hfg, val_aux_zero], exact zero_le _ },
+  replace hf : ∃ n, coeff _ _ n f ≠ 0 := not_forall.1 (λ h, hf $ perfection.ext h),
+  replace hg : ∃ n, coeff _ _ n g ≠ 0 := not_forall.1 (λ h, hg $ perfection.ext h),
+  replace hfg : ∃ n, coeff _ _ n (f + g) ≠ 0 := not_forall.1 (λ h, hfg $ perfection.ext h),
+  obtain ⟨m, hm⟩ := hf, obtain ⟨n, hn⟩ := hg, obtain ⟨k, hk⟩ := hfg,
+  replace hm := coeff_ne_zero_of_le hm (le_trans (le_max_left m n) (le_max_left _ k)),
+  replace hn := coeff_ne_zero_of_le hn (le_trans (le_max_right m n) (le_max_left _ k)),
+  replace hk := coeff_ne_zero_of_le hk (le_max_right (max m n) k),
+  rw [val_aux_eq hm, val_aux_eq hn, val_aux_eq hk, ring_hom.map_add],
+  cases le_max_iff.1
+    (mod_p.pre_val_add (coeff _ _ (max (max m n) k) f) (coeff _ _ (max (max m n) k) g)) with h h,
+  { exact le_max_left_of_le (canonically_ordered_semiring.pow_le_pow_of_le_left h _) },
+  { exact le_max_right_of_le (canonically_ordered_semiring.pow_le_pow_of_le_left h _) }
+end
+
+variables (K v O hv p)
+/-- The valuation `Perfection(O/(p)) → ℝ≥0`.
+Given `f ∈ Perfection(O/(p))`, if `f = 0` then output `0`;
+otherwise output `pre_val(f(n))^(p^n)` for any `n` such that `f(n) ≠ 0`. -/
+noncomputable def val : valuation (pre_tilt K v O hv p) ℝ≥0 :=
+{ to_fun := val_aux K v O hv p,
+  map_one' := val_aux_one,
+  map_mul' := val_aux_mul,
+  map_zero' := val_aux_zero,
+  map_add' := val_aux_add }
+
+variables {K v O hv p}
+lemma map_eq_zero {f : pre_tilt K v O hv p} : val K v O hv p f = 0 ↔ f = 0 :=
+begin
+  by_cases hf0 : f = 0, { rw hf0, exact iff_of_true (valuation.map_zero _) rfl },
+  obtain ⟨n, hn⟩ : ∃ n, coeff _ _ n f ≠ 0 := not_forall.1 (λ h, hf0 $ perfection.ext h),
+  show val_aux K v O hv p f = 0 ↔ f = 0, refine iff_of_false (λ hvf, hn _) hf0,
+  rw val_aux_eq hn at hvf, replace hvf := nnreal.pow_eq_zero hvf, rwa mod_p.pre_val_eq_zero at hvf
+end
+
+end classical
+
+instance : integral_domain (pre_tilt K v O hv p) :=
+{ exists_pair_ne := (char_p.nontrivial_of_char_ne_one hp.ne_one).1,
+  eq_zero_or_eq_zero_of_mul_eq_zero := λ f g hfg,
+    by { simp_rw ← map_eq_zero at hfg ⊢, contrapose! hfg, rw valuation.map_mul,
+      exact nnreal.mul_ne_zero' hfg.1 hfg.2 },
+  .. (infer_instance : comm_ring (pre_tilt K v O hv p)) }
+
+end pre_tilt
+
+/-- The tilt of a field, as defined in Perfectoid Spaces by Peter Scholze.
```

Comment:
Please add a reference to the `.bib` file.

Reply:
Done